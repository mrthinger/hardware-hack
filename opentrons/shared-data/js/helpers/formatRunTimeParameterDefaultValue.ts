import type { ValueRunTimeParameter } from '../types'

/**
 * Formats the runtime parameter's default value.
 *
 * @param {RunTimeParameter} runTimeParameter - The runtime parameter whose default value is to be formatted.
 * @param {Function} [t] - An optional function for localization.
 *
 * @returns {string} The formatted default value of the runtime parameter.
 *
 */

export const formatRunTimeParameterDefaultValue = (
  runTimeParameter: ValueRunTimeParameter,
  t?: any
): string => {
  const { type, default: defaultValue } = runTimeParameter
  const suffix =
    'suffix' in runTimeParameter && runTimeParameter.suffix != null
      ? runTimeParameter.suffix
      : null

  if ('choices' in runTimeParameter && runTimeParameter.choices != null) {
    const choice = runTimeParameter.choices.find(
      choice => choice.value === defaultValue
    )
    if (choice != null) {
      return suffix != null
        ? `${choice.displayName} ${suffix}`
        : choice.displayName
    }
  }

  switch (type) {
    case 'int':
    case 'float':
      return suffix !== null
        ? `${defaultValue.toString()} ${suffix}`
        : defaultValue.toString()
    case 'bool':
      if (t != null) {
        return Boolean(defaultValue) ? t('on') : t('off')
      } else {
        return Boolean(defaultValue) ? 'On' : 'Off'
      }
    default:
      break
  }
  return ''
}
