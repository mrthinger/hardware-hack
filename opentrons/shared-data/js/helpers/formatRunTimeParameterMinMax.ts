import type { RunTimeParameter } from '../types'
/**
 * Formats the runtime parameter's minimum and maximum values.
 *
 * @param {RunTimeParameter} runTimeParameter - The runtime parameter whose min and max values are to be formatted.
 *
 * @returns {string} The formatted min-max value of the runtime parameter.
 *
 * @example
 * const runTimeParameter = {
 *    value: 6.5,
 *   displayName: 'EtoH Volume',
 *     variableName: 'ETOH_VOLUME',
 *     description: '70% ethanol volume',
 *     type: 'float',
 *     suffix: 'mL',
 *     min: 1.5,
 *     max: 10.0,
 *     default: 6.5,
 * }
 * console.log(formatRunTimeParameterMinMax(runTimeParameter)); // "1.5-10.0"
 */

export const formatRunTimeParameterMinMax = (
  runTimeParameter: RunTimeParameter
): string => {
  const min = 'min' in runTimeParameter ? runTimeParameter.min : 0
  const max = 'max' in runTimeParameter ? runTimeParameter.max : 0
  return runTimeParameter.type === 'int'
    ? `${min}-${max}`
    : `${min.toFixed(1)}-${max.toFixed(1)}`
}
