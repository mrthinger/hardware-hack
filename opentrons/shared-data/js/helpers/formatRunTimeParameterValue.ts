import type { RunTimeParameter } from '../types'

/**
 * Formats the runtime parameter value.
 *
 * @param {ValueRunTimeParameter} runTimeParameter - The value runtime parameter to be formatted.
 * @param {Function} t - A function for localization.
 *
 * @returns {string} The formatted runtime parameter value.
 *
 */

export const formatRunTimeParameterValue = (
  runTimeParameter: RunTimeParameter,
  t: any
): string => {
  const { type } = runTimeParameter
  const value =
    runTimeParameter.type === 'csv_file'
      ? runTimeParameter.file?.file?.name ?? ''
      : runTimeParameter.value
  const suffix =
    'suffix' in runTimeParameter && runTimeParameter.suffix != null
      ? runTimeParameter.suffix
      : null

  if ('choices' in runTimeParameter && runTimeParameter.choices != null) {
    const choice = runTimeParameter.choices.find(
      choice => choice.value === value
    )
    if (choice != null) {
      return suffix != null
        ? `${choice.displayName} ${suffix}`
        : choice.displayName
    }
  }
  switch (type) {
    case 'int':
    case 'float':
      return suffix !== null
        ? `${value.toString()} ${suffix}`
        : value.toString()
    case 'bool': {
      return Boolean(value) ? t('on') : t('off')
    }
    case 'csv_file':
      return value.toString()
    default:
      break
  }
  return ''
}
