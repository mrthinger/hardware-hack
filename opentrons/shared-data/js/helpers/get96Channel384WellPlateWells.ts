//  TODO(jr, 10/19/23): should extend getWellNamePerMultiTip to use similar math for 96-channel and 384 well plates
//  instead of special casing using this util
export function get96Channel384WellPlateWells(
  all384Wells: string[],
  well: string
): string[] {
  const totalWells = 96 * 2 //  multiplying 2 because we will remove either odd or even numbers

  const filterWellsFor96Channel = (start: number, isOdd: boolean): string[] => {
    const filteredWells: string[] = []
    let count = start
    for (let i = 0; i < totalWells; i++) {
      if (count < all384Wells.length) {
        const well = all384Wells[count]
        const numberFromWell = well.match(/\d+/)
        if (numberFromWell) {
          const number = parseInt(numberFromWell[0])
          if ((number % 2 === 1 && isOdd) || (number % 2 === 0 && !isOdd)) {
            filteredWells.push(well)
          }
        }
        count += 2
      } else {
        break
      }
    }
    return filteredWells
  }

  const allSetsOfWells = [
    filterWellsFor96Channel(0, true),
    filterWellsFor96Channel(1, true),
    filterWellsFor96Channel(0, false),
    filterWellsFor96Channel(1, false),
  ]

  for (const wells of allSetsOfWells) {
    if (wells.includes(well)) {
      return wells
    }
  }

  return []
}
