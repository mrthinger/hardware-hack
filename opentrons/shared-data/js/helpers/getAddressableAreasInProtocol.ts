import { MOVABLE_TRASH_A3_ADDRESSABLE_AREA } from '../constants'
import {
  getAddressableAreaNamesFromLoadedModule,
  getAddressableAreaFromSlotId,
} from '../fixtures'
import type { AddressableAreaName } from '../../deck'
import type { ProtocolAnalysisOutput } from '../../protocol'
import type { CompletedProtocolAnalysis, DeckDefinition } from '../types'

export function getAddressableAreasInProtocol(
  protocolAnalysis: CompletedProtocolAnalysis | ProtocolAnalysisOutput,
  deckDef: DeckDefinition
): AddressableAreaName[] {
  const { commands, labware } = protocolAnalysis

  const addressableAreasFromCommands = commands.reduce<AddressableAreaName[]>(
    (acc, command) => {
      const { commandType, params } = command
      if (
        commandType === 'moveLabware' &&
        params.newLocation !== 'offDeck' &&
        'slotName' in params.newLocation &&
        !acc.includes(params.newLocation.slotName as AddressableAreaName)
      ) {
        const addressableAreaName = getAddressableAreaFromSlotId(
          params.newLocation.slotName,
          deckDef
        )?.id

        if (addressableAreaName == null) {
          return acc
        } else {
          return [...acc, addressableAreaName]
        }
      } else if (
        commandType === 'moveLabware' &&
        params.newLocation !== 'offDeck' &&
        'addressableAreaName' in params.newLocation &&
        !acc.includes(params.newLocation.addressableAreaName)
      ) {
        return [...acc, params.newLocation.addressableAreaName]
      } else if (
        commandType === 'loadLabware' &&
        params.location !== 'offDeck' &&
        'slotName' in params.location &&
        !acc.includes(params.location.slotName as AddressableAreaName)
      ) {
        const addressableAreaName = getAddressableAreaFromSlotId(
          params.location.slotName,
          deckDef
        )?.id

        // do not add addressable area name for legacy trash labware
        if (
          addressableAreaName == null ||
          ('loadName' in params &&
            params.loadName === 'opentrons_1_trash_3200ml_fixed')
        ) {
          return acc
        } else {
          return [...acc, addressableAreaName]
        }
      } else if (
        commandType === 'loadModule' &&
        !acc.includes(params.location.slotName as AddressableAreaName)
      ) {
        const addressableAreaNames = getAddressableAreaNamesFromLoadedModule(
          params.model,
          params.location.slotName,
          deckDef
        )

        return [...acc, ...addressableAreaNames]
      } else if (
        commandType === 'loadLabware' &&
        params.location !== 'offDeck' &&
        'addressableAreaName' in params.location &&
        !acc.includes(params.location.addressableAreaName)
      ) {
        return [...acc, params.location.addressableAreaName]
      } else if (
        commandType === 'moveToAddressableArea' &&
        !acc.includes(params.addressableAreaName)
      ) {
        return [...acc, params.addressableAreaName]
      } else if (
        commandType === 'moveToAddressableAreaForDropTip' &&
        !acc.includes(params.addressableAreaName)
      ) {
        return [...acc, params.addressableAreaName]
      } else {
        return acc
      }
    },
    []
  )

  // special-case the Flex trash labware load name in A3 for back compatibility with the legacy fixed trash load labware command
  const legacyTrashAddressableArea = labware.some(
    ({ loadName, location }) =>
      loadName === 'opentrons_1_trash_3200ml_fixed' &&
      location !== 'offDeck' &&
      'slotName' in location &&
      location.slotName === 'A3'
  )
    ? MOVABLE_TRASH_A3_ADDRESSABLE_AREA
    : []

  return addressableAreasFromCommands.concat(legacyTrashAddressableArea)
}
