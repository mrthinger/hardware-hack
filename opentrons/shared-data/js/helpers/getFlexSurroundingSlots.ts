import type { DeckSlotId } from '../types'

export const FLEX_GRID = [
  ['A1', 'A2', 'A3'],
  ['B1', 'B2', 'B3'],
  ['C1', 'C2', 'C3'],
  ['D1', 'D2', 'D3'],
]

const LETTER_TO_ROW_MAP: Record<string, number> = {
  A: 0,
  B: 1,
  C: 2,
  D: 3,
}

let COLS = 3 // Initial number of columns in each row
const ROWS = 4

const DIRECTIONS = [
  [-1, -1], // NW
  [-1, 0], // N
  [-1, 1], // NE
  [0, -1], // W
  [0, 1], // E
  [1, -1], // SW
  [1, 0], // S
  [1, 1], // SE
]

export const getFlexSurroundingSlots = (
  slot: DeckSlotId,
  stagingAreaSlots: DeckSlotId[]
): DeckSlotId[] => {
  // Handle staging area slots
  if (stagingAreaSlots.length > 0) {
    stagingAreaSlots.forEach((stagingSlot, index) => {
      if (stagingSlot) {
        FLEX_GRID[index].push(stagingSlot)
      }
    })
    COLS = Math.max(COLS, FLEX_GRID[0].length) // Update COLS to the maximum row length
  }

  const letter = slot.charAt(0)
  const col = parseInt(slot.charAt(1)) - 1 // Convert the column to a 0-based index
  const row = LETTER_TO_ROW_MAP[letter]

  const surroundingSlots: DeckSlotId[] = []

  // Iterate through both directions
  DIRECTIONS.forEach(([dRow, dCol]) => {
    const newRow = row + dRow
    const newCol = col + dCol

    if (newRow >= 0 && newRow < ROWS && newCol >= 0 && newCol < COLS) {
      surroundingSlots.push(FLEX_GRID[newRow][newCol])
    }
  })

  // Filter out any undefined values from the staging area slots that are not added
  return surroundingSlots.filter(slot => slot !== undefined)
}
