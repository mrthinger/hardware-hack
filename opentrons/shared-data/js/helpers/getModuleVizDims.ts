// NOTE: Ian 2019-10-24 these are by-eye numbers that intentionally
// ignore the real-life data to emphasize overhangs etc
import { SPAN7_8_10_11_SLOT } from '../constants'
import type { ModuleOrientation } from '../types'
// NOTE: all dims are in 'left' orientation. Rotate & transform to obtain 'right' orientation.

export interface ModuleVizDims {
  xOffset: number
  yOffset: number
  xDimension: number
  yDimension: number
  childXOffset: number
  childYOffset: number
  childXDimension: number
  childYDimension: number
}
type DeckSlot = string
const LEFT_SIDE_SLOTS: DeckSlot[] = ['1', '4', '7', '10', SPAN7_8_10_11_SLOT]
export const inferModuleOrientationFromSlot = (
  slot: DeckSlot
): ModuleOrientation => {
  // NOTE: defaults to 'left' even for center slots and spanning slots.
  // That's consistent with non-rotatable modules (thermocycler) being
  // considered as 'left' all the time, like we do in MODULE_VIZ_DIMS.
  if (!LEFT_SIDE_SLOTS.includes(slot)) {
    return 'right'
  }

  return 'left'
}

export const inferModuleOrientationFromXCoordinate = (
  xCoordinate: number
): ModuleOrientation => {
  if (xCoordinate === 0) {
    return 'left'
  }
  return 'right'
}
