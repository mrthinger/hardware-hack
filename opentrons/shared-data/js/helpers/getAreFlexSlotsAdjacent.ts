import { FLEX_GRID } from './getFlexSurroundingSlots'
import type { DeckSlotId } from '../types'

export const getAreFlexSlotsAdjacent = (
  slot1: DeckSlotId,
  slot2: DeckSlotId
): boolean => {
  const findSlotPosition = (slot: DeckSlotId): [number, number] | null => {
    for (let row = 0; row < FLEX_GRID.length; row++) {
      const col = FLEX_GRID[row].indexOf(slot)
      if (col !== -1) {
        return [row, col]
      }
    }
    return null
  }

  const pos1 = findSlotPosition(slot1)
  const pos2 = findSlotPosition(slot2)

  if (pos1 === null || pos2 === null) {
    return false
  }

  const [row1, col1] = pos1
  const [row2, col2] = pos2

  const rowDiff = Math.abs(row1 - row2)
  const colDiff = Math.abs(col1 - col2)

  if ((rowDiff === 1 && colDiff <= 1) || (rowDiff === 0 && colDiff === 1)) {
    return true
  }

  return false
}
