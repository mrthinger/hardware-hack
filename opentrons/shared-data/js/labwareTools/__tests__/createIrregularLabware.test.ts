import omit from 'lodash/omit'
import range from 'lodash/range'
import { describe, it, expect, beforeEach } from 'vitest'
import { splitWellsOnColumn, sortWells } from '../../helpers/index'
import fixture_irregular_example_1 from '../../../labware/fixtures/2/fixture_irregular_example_1.json'

import {
  createIrregularLabware,
  _irregularWellName,
  _generateIrregularLoadName,
  _calculateWellCoord,
} from '..'

import type { LabwareDefinition2, LabwareWellProperties } from '../../types'
import type { IrregularLabwareProps } from '..'

// NOTE: loadName needs to be replaced here b/c fixture has a non-default loadName
const exampleLabware1 = {
  ...fixture_irregular_example_1,
  parameters: {
    ...fixture_irregular_example_1.parameters,
    loadName: 'generic_55_tuberack_50x3ml_5x10ml',
  },
} as LabwareDefinition2

describe('test helper functions', () => {
  it('Well name generated correctly', () => {
    const grid = { row: 2, column: 2 }

    const gridStart = [
      { rowStart: 'A', colStart: '1', rowStride: 1, colStride: 2 },
      { rowStart: 'B', colStart: '1', rowStride: 3, colStride: 1 },
    ]
    const expected1 = ['A1', 'B1', 'A3', 'B3']
    const expected2 = ['B1', 'E1', 'B2', 'E2']

    range(grid.column).forEach(colIdx => {
      range(grid.row).forEach(rowIdx => {
        const idx = colIdx * grid.row + rowIdx

        const wellName1 = _irregularWellName(rowIdx, colIdx, gridStart[0])

        expect(expected1[idx]).toEqual(wellName1)

        const wellName2 = _irregularWellName(rowIdx, colIdx, gridStart[1])

        expect(expected2[idx]).toEqual(wellName2)
      })
    })
  })

  it('XYZ generates correctly for each grid', () => {
    const grid = { row: 1, column: 5 }
    const offset = { x: 1, y: 0.5, z: 55.5 }
    const spacing = [
      { row: 10, column: 10 },
      { row: 5, column: 14 },
    ]
    const well = [
      omit(exampleLabware1.wells.A1, ['x', 'y', 'z']),
      omit(exampleLabware1.wells.B1, ['x', 'y', 'z']),
    ] as LabwareWellProperties[]

    const expectedX1 = [1, 11, 21, 31, 41]
    const expectedY1 = [0.5]
    const expectedX2 = [1, 15, 29, 43, 57]
    const expectedY2 = [0.5]

    range(grid.column).forEach(colIdx => {
      range(grid.row).forEach(rowIdx => {
        const well1 = _calculateWellCoord(
          rowIdx,
          colIdx,
          spacing[0],
          offset,
          well[0]
        )

        expect(well1.x).toBeCloseTo(expectedX1[colIdx], 2)
        expect(well1.y).toBeCloseTo(expectedY1[rowIdx], 2)
        expect(well1.z).toBeCloseTo(offset.z - well[0].depth, 2)

        const well2 = _calculateWellCoord(
          rowIdx,
          colIdx,
          spacing[1],
          offset,
          well[1]
        )

        expect(well2.x).toBeCloseTo(expectedX2[colIdx], 2)
        expect(well2.y).toBeCloseTo(expectedY2[rowIdx], 2)
        expect(well2.z).toBeCloseTo(offset.z - well[1].depth, 2)
      })
    })
  })
})

describe('test createIrregularLabware function', () => {
  let labware1: LabwareDefinition2
  let labware1Args: IrregularLabwareProps

  beforeEach(() => {
    labware1Args = {
      namespace: 'fixture',
      metadata: {
        displayName: 'Fake Irregular Container',
        displayCategory: 'tubeRack',
        displayVolumeUnits: 'mL',
        tags: ['fake', 'opentrons'],
      },
      parameters: {
        format: 'irregular',
        isTiprack: false,
        isMagneticModuleCompatible: false,
      },
      dimensions: {
        xDimension: 127.76,
        yDimension: 85.48,
        zDimension: 64.48,
      },
      well: [
        {
          depth: 10.54,
          shape: 'circular',
          diameter: 10,
          totalLiquidVolume: 3000,
        },
        {
          depth: 20.54,
          shape: 'circular',
          diameter: 15,
          totalLiquidVolume: 10000,
        },
      ],
      offset: [
        { x: 10, y: 10, z: 69.48 },
        { x: 15, y: 15, z: 69.48 },
      ],
      grid: [
        { row: 5, column: 10 },
        { row: 1, column: 5 },
      ],
      spacing: [
        { row: 5, column: 10 },
        { row: 5, column: 10 },
      ],
      gridStart: [
        { rowStart: 'A', colStart: '1', rowStride: 2, colStride: 1 },
        { rowStart: 'B', colStart: '1', rowStride: 1, colStride: 1 },
      ],
    }
    labware1 = createIrregularLabware(labware1Args)
  })

  it('irregular ordering generates as expected', () => {
    const keyList = Object.keys(labware1.wells)
    const generatedOrdering = splitWellsOnColumn(keyList.sort(sortWells))
    expect(labware1.ordering).toEqual(generatedOrdering)
  })

  it('check labware matches fixture', () => {
    expect(labware1).toEqual(exampleLabware1)
  })

  it('labware loadName generated correctly for multi-grid labware', () => {
    const loadName = _generateIrregularLoadName({
      grid: [
        { row: 3, column: 2 },
        { row: 1, column: 4 },
      ],
      well: [
        {
          depth: 20,
          shape: 'circular',
          totalLiquidVolume: 400,
        } as LabwareWellProperties,
        {
          depth: 110,
          shape: 'circular',
          totalLiquidVolume: 2000,
        } as LabwareWellProperties,
      ],
      totalWellCount: 10,
      units: 'ÂµL',
      displayCategory: 'wellPlate',
      brand: 'some brand',
    })

    expect(loadName).toEqual('somebrand_10_wellplate_6x400ul_4x2000ul')
  })

  it('labware loadName generated correctly for multi-grid labware in other units', () => {
    const loadName = _generateIrregularLoadName({
      grid: [{ row: 3, column: 2 }],
      well: [
        {
          depth: 20,
          shape: 'circular',
          totalLiquidVolume: 4000,
        } as LabwareWellProperties,
      ],
      totalWellCount: 6,
      units: 'mL',
      displayCategory: 'wellPlate',
      brand: 'some brand',
    })

    expect(loadName).toEqual('somebrand_6_wellplate_6x4ml')
  })

  it('failing to validate against labware schema throws w/o "strict"', () => {
    const args = {
      ...labware1Args,
      // negative y offset should fail schema validation by making well `y` negative
      offset: [
        { x: 10, y: -9999, z: 69.48 },
        { x: 15, y: -9999, z: 69.48 },
      ],
    }
    expect(() => createIrregularLabware(args)).toThrowErrorMatchingSnapshot()
    expect(() =>
      createIrregularLabware({ ...args, strict: false })
    ).not.toThrow()
  })
})
