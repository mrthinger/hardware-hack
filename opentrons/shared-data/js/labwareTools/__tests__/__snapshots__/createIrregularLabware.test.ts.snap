// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`test createIrregularLabware function > failing to validate against labware schema throws w/o "strict" 1`] = `[Error: Generated labware failed to validate, please check your inputs]`;

exports[`test createIrregularLabware function failing to validate against labware schema throws w/o "strict" 1`] = `"Generated labware failed to validate, please check your inputs"`;
